File: src-tauri/resources/cat-ptt/catptt_bridge.py
#!/usr/bin/env python3
"""Close-serial CAT-PTT bridge for radios that key by CAT command only.

A radio such as the Yaesu FT-710 keys ONLY by CAT command (TX1; / TX0;) on its
serial port; serial RTS/DTR PTT is a no-op. On a single-cable USB tree (hub ->
CP2105 serial + C-Media codec) the codec resets if the serial port is held OPEN
concurrent with audio streaming. The fix proven on air 2026-06-23: key by CAT,
then CLOSE the serial port during audio (the radio stays CAT-latched in TX),
reopening only momentarily to relay each keystring.

ardopcf is launched with `-c TCP:<port> -k <hex(key)> -u <hex(unkey)>` so it
sends its keystring over this TCP socket instead of toggling a serial line. This
bridge listens on that socket, and for each keystring opens the serial port,
writes the bytes, flushes, and closes — momentary open/write/close — so the port
is shut while the codec streams audio.

Parameters are read from argv (see `--help`) so tuxlink can drive any radio in
this class without editing the script. The defaults reproduce the FT-710 setup.

On every connection teardown the bridge sends the unkey command as a failsafe so
a dropped ardopcf socket cannot leave the radio latched in TX.
"""

import argparse
import signal
import socket
import sys
import time

try:
    import serial  # pyserial
except ImportError:
    sys.stderr.write(
        "[bridge] FATAL: pyserial not installed (python3 -m pip install pyserial)\n"
    )
    sys.exit(2)


def parse_args(argv):
    p = argparse.ArgumentParser(description="Close-serial CAT-PTT bridge.")
    p.add_argument("--port", type=int, default=4532,
                   help="TCP port to listen on (loopback). Default 4532.")
    p.add_argument("--serial", default="/dev/ttyUSB0",
                   help="Serial device for CAT. Default /dev/ttyUSB0.")
    p.add_argument("--baud", type=int, default=38400,
                   help="Serial baud rate. Default 38400.")
    p.add_argument("--key", default="TX1;",
                   help="CAT key command. Default TX1;.")
    p.add_argument("--unkey", default="TX0;",
                   help="CAT unkey command. Default TX0;.")
    return p.parse_args(argv)


def main(argv):
    args = parse_args(argv)
    key_bytes = args.key.encode("ascii")
    unkey_bytes = args.unkey.encode("ascii")
    maxlen = max(len(key_bytes), len(unkey_bytes))

    def cat_send(payload, label):
        """Momentary open/write/close so the serial port is shut during audio."""
        try:
            ser = serial.Serial(args.serial, args.baud, timeout=0.4,
                                rtscts=False, dsrdtr=False)
            ser.rts = False
            ser.dtr = False
            ser.write(payload)
            ser.flush()
            time.sleep(0.07)
            ser.close()
            print(f"[bridge] >>> {label}", flush=True)
        except Exception as exc:  # noqa: BLE001 — relay best-effort, never crash
            print(f"[bridge] CAT err {exc!r}", flush=True)

    # Failsafe on termination: tuxlink stops the bridge with SIGINT (escalating
    # to SIGKILL). Send the unkey command on SIGINT/SIGTERM so a teardown while
    # the radio is latched in TX cannot leave it keyed. SIGKILL cannot be caught,
    # but tuxlink only escalates to it after a grace period in which this handler
    # has already run.
    def _failsafe(_signum, _frame):
        cat_send(unkey_bytes, "UNKEY (signal failsafe)")
        sys.exit(0)

    signal.signal(signal.SIGINT, _failsafe)
    signal.signal(signal.SIGTERM, _failsafe)

    srv = socket.socket()
    srv.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    srv.bind(("127.0.0.1", args.port))
    srv.listen(1)
    print(f"[bridge] listening 127.0.0.1:{args.port} -> {args.serial}@{args.baud} "
          f"key={args.key!r} unkey={args.unkey!r}", flush=True)

    while True:
        conn, _ = srv.accept()
        print("[bridge] ardopcf connected", flush=True)
        buf = b""
        try:
            while True:
                data = conn.recv(256)
                if not data:
                    break
                buf += data
                print(f"[bridge rx] {data!r}", flush=True)
                # Match the FULL configured key/unkey byte sequences (e.g. b"TX1;"
                # / b"TX0;"), not a ';'-delimited substring. This processes
                # commands that carry no ';' terminator, and avoids
                # overlapping-prefix misrouting (e.g. key "TX" vs unkey "TX0"). On
                # any ambiguity the UNKEY wins, so the radio is never accidentally
                # left keyed.
                while True:
                    ki = buf.find(key_bytes)
                    ui = buf.find(unkey_bytes)
                    if ki < 0 and ui < 0:
                        break
                    if ui >= 0 and (ki < 0 or ui <= ki):
                        cat_send(unkey_bytes, "UNKEY")
                        buf = buf[ui + len(unkey_bytes):]
                    else:
                        cat_send(key_bytes, "KEY")
                        buf = buf[ki + len(key_bytes):]
                # No complete command remains. Keep only a tail that could be the
                # prefix of a future full match so unrecognized noise can't grow
                # the buffer without bound.
                if len(buf) >= maxlen:
                    buf = buf[-(maxlen - 1):] if maxlen > 1 else b""
        except Exception as exc:  # noqa: BLE001
            print(f"[bridge] conn end {exc!r}", flush=True)
        finally:
            # Failsafe: never leave the radio latched in TX on a dropped socket.
            cat_send(unkey_bytes, "UNKEY (failsafe)")
            conn.close()


if __name__ == "__main__":
    main(sys.argv[1:])

File: src-tauri/src/winlink/modem/ardop/cat_ptt_bridge.rs
//! Close-serial CAT-PTT bridge (tuxlink-wu0k).
//!
//! Some radios — notably the Yaesu FT-710 — key ONLY by a CAT command
//! (`TX1;` / `TX0;`), and on a single-cable USB tree (hub → CP2105 serial +
//! C-Media codec) the audio codec RESETS if the serial port is held OPEN while
//! audio streams. The fix proven on air 2026-06-23 (FT-710 + Pi 5): key by CAT,
//! then CLOSE the serial port during audio — the radio stays CAT-latched in TX
//! — reopening only momentarily to relay each keystring.
//!
//! ardopcf is spawned with `-c TCP:<port> -k <hex(key)> -u <hex(unkey)>` (see
//! `crate::modem_commands::build_ardop_extra_args`), which makes it send its
//! keystring over a TCP "CAT" socket instead of toggling a serial RTS line.
//! This module owns the process that serves that socket: a tiny, audited Python
//! script (`resources/cat-ptt/catptt_bridge.py`) that does the momentary
//! open/write/close. The script is embedded in the binary via `include_str!`
//! and run with the system `python3`, so the bridge needs no extra Rust crate
//! (no cold-compile risk) and behaves identically in a dev run and a packaged
//! build (no Tauri resource-path resolution to get wrong).
//!
//! # Lifecycle
//!
//! [`CatPttBridge::spawn`] writes the embedded script to a private temp file and
//! launches `python3 <script> --port … --serial … --baud … --key … --unkey …`
//! under a [`ManagedModem`] supervisor, then waits for the loopback port to
//! accept connections. The caller
//! ([`super::transport::ArdopTransport::with_managed_modem`]) starts the bridge
//! BEFORE ardopcf and stops it via [`CatPttBridge::stop`] on shutdown; `Drop`
//! is a backstop. The Python script sends the unkey command on every connection
//! teardown as its own failsafe, so a dropped ardopcf socket cannot leave the
//! radio latched in TX.
//!
//! # RADIO-1
//!
//! Spawning the bridge does NOT key the radio: ardopcf only emits a keystring
//! after a CONNECT, which is behind the existing consent gate. The bridge only
//! relays what ardopcf sends. No tuxlink-added airtime cap or timer lives here.

use std::io::{self, Write};
use std::net::TcpStream;
use std::path::PathBuf;
use std::time::{Duration, Instant};

use super::super::process::ManagedModem;

/// The embedded close-serial CAT-PTT bridge script. Tracked at
/// `src-tauri/resources/cat-ptt/catptt_bridge.py`; embedded so the bridge runs
/// identically in dev and packaged builds without Tauri resource-path lookups.
const BRIDGE_SCRIPT: &str = include_str!("../../../../resources/cat-ptt/catptt_bridge.py");

/// Total time to wait for the bridge to bind its loopback port after spawn.
const BIND_WAIT: Duration = Duration::from_secs(5);
/// Poll interval while waiting for the bridge port to come up.
const BIND_POLL: Duration = Duration::from_millis(50);

/// Parameters for the close-serial CAT-PTT bridge (tuxlink-wu0k). Built from
/// [`crate::config::ArdopUiConfig`] when `ptt_method == CatCommand`.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct CatBridgeSpec {
    /// Loopback TCP port the bridge listens on; ardopcf connects via
    /// `-c TCP:<port>`. Distinct from the ARDOP cmd/data ports.
    pub bridge_port: u16,
    /// Serial device for CAT (e.g. `/dev/ttyUSB0`).
    pub serial_path: String,
    /// Serial baud rate (e.g. 38400 for the FT-710 Enhanced port).
    pub baud: u32,
    /// CAT key command (e.g. `TX1;`).
    pub key_cmd: String,
    /// CAT unkey command (e.g. `TX0;`).
    pub unkey_cmd: String,
}

impl CatBridgeSpec {
    /// Build the `python3` argv for this spec (after the script path). Pure, so
    /// it is unit-testable without spawning anything.
    pub fn python_args(&self, script_path: &str) -> Vec<String> {
        vec![
            script_path.to_string(),
            "--port".to_string(),
            self.bridge_port.to_string(),
            "--serial".to_string(),
            self.serial_path.clone(),
            "--baud".to_string(),
            self.baud.to_string(),
            "--key".to_string(),
            self.key_cmd.clone(),
            "--unkey".to_string(),
            self.unkey_cmd.clone(),
        ]
    }
}

/// A running CAT-PTT bridge process plus the temp script file it runs from.
///
/// Owns a [`ManagedModem`] (the supervised `python3` process) and the temp
/// script path. `stop` terminates the process; `Drop` removes the temp file and
/// is a backstop terminator via `ManagedModem`'s own `Drop`.
#[derive(Debug)]
pub struct CatPttBridge {
    process: ManagedModem,
    script_path: PathBuf,
    bridge_port: u16,
}

impl CatPttBridge {
    /// Write the embedded script to a private temp file, spawn
    /// `python3 <script> …`, and wait for the loopback port to accept.
    ///
    /// Returns an error if the temp file cannot be written, `python3` cannot be
    /// spawned, or the port never binds within [`BIND_WAIT`].
    pub fn spawn(spec: &CatBridgeSpec) -> io::Result<CatPttBridge> {
        // Port-collision pre-check, BEFORE any side effects. The default bridge
        // port (4532) is hamlib rigctld's upstream default; tuxlink defaults its
        // own rigctld to 4534 to avoid this exact collision (C1 fix,
        // tuxlink-8fkkk), but a manually-launched rigctld or any other process
        // may still hold 4532. If another process already holds the port, the
        // post-spawn TCP probe below would accept against THAT listener — and
        // ardopcf would key the wrong process while our bridge silently failed
        // to bind, with its unkey failsafe out of the path. Fail loudly instead.
        // The listener is dropped immediately so the bridge child can bind the
        // port within milliseconds.
        match std::net::TcpListener::bind(("127.0.0.1", spec.bridge_port)) {
            Ok(listener) => drop(listener),
            Err(e) => {
                return Err(io::Error::new(
                    io::ErrorKind::AddrInUse,
                    format!(
                        "CAT-PTT bridge port {} is unavailable ({e}) — rigctld \
                         defaults to 4532, or another session holds it; choose a \
                         different CAT bridge port",
                        spec.bridge_port
                    ),
                ));
            }
        }

        let script_path = write_bridge_script()?;
        let script_str = script_path.to_string_lossy().into_owned();

        let args = spec.python_args(&script_str);
        let arg_refs: Vec<&str> = args.iter().map(String::as_str).collect();

        let process = ManagedModem::spawn("python3", &arg_refs).map_err(|e| {
            // Best-effort cleanup of the temp script if the spawn failed.
            let _ = std::fs::remove_file(&script_path);
            io::Error::other(format!("failed to spawn CAT-PTT bridge (python3): {e}"))
        })?;

        let mut bridge = CatPttBridge {
            process,
            script_path,
            bridge_port: spec.bridge_port,
        };

        // Wait for the loopback port to come up so ardopcf's `-c TCP:<port>`
        // connect (which follows immediately) does not race the bind.
        bridge.wait_for_port(BIND_WAIT)?;
        Ok(bridge)
    }

    /// The loopback port the bridge listens on.
    pub fn bridge_port(&self) -> u16 {
        self.bridge_port
    }

    /// Poll `TcpStream::connect` to the bridge port until it accepts or the
    /// timeout elapses. Each successful probe connection is immediately closed.
    ///
    /// Also checks the spawned child is still alive on each poll: if the bridge
    /// process exited during startup (e.g. a serial-open or late bind failure),
    /// fail immediately rather than wait out the timeout — and never report a
    /// dead bridge as ready.
    fn wait_for_port(&mut self, timeout: Duration) -> io::Result<()> {
        let addr = format!("127.0.0.1:{}", self.bridge_port);
        let start = Instant::now();
        loop {
            if !self.process.is_running() {
                return Err(io::Error::other(format!(
                    "CAT-PTT bridge process exited during startup (port {}) — \
                     check the serial device and that the port is free",
                    self.bridge_port
                )));
            }
            if TcpStream::connect_timeout(
                &addr.parse().map_err(|e| {
                    io::Error::new(io::ErrorKind::InvalidInput, format!("bad bridge addr: {e}"))
                })?,
                BIND_POLL,
            )
            .is_ok()
            {
                return Ok(());
            }
            if start.elapsed() >= timeout {
                return Err(io::Error::new(
                    io::ErrorKind::TimedOut,
                    format!(
                        "CAT-PTT bridge did not bind port {} within {timeout:?}",
                        self.bridge_port
                    ),
                ));
            }
            std::thread::sleep(BIND_POLL);
        }
    }

    /// Stop the bridge process gracefully (SIGINT → grace → SIGKILL) and remove
    /// the temp script file. Idempotent. On stop the Python script's connection
    /// teardown failsafe has already sent the unkey command for any open socket.
    pub fn stop(&mut self, grace: Duration) -> io::Result<()> {
        let result = self
            .process
            .stop(grace)
            .map_err(|e| io::Error::other(format!("CAT-PTT bridge stop failed: {e}")));
        // Remove the temp script regardless of stop outcome.
        let _ = std::fs::remove_file(&self.script_path);
        result
    }
}

impl Drop for CatPttBridge {
    fn drop(&mut self) {
        // The real Tauri disconnect path DROPS the transport (it holds it as a
        // `Box<dyn ModemTransport>` and `modem_ardop_disconnect_inner` drops it
        // rather than calling `ArdopTransport::shutdown`), so this Drop — not
        // `shutdown` — is the usual bridge teardown. Give the bridge a real grace
        // window to run its SIGINT unkey-failsafe (open serial → write TX0; →
        // close) instead of relying on `ManagedModem::Drop`'s short 200 ms
        // SIGINT→SIGKILL escalation, which can cut the unkey off mid-write and
        // strand the radio keyed. Best-effort; `ManagedModem::Drop` terminates the
        // process regardless. Idempotent with `stop` on the `shutdown` path.
        let _ = self.process.stop(Duration::from_secs(3));
        // Remove the temp script so we don't leak files across reconnects.
        let _ = std::fs::remove_file(&self.script_path);
    }
}

/// Write the embedded bridge script to a uniquely-named temp file and return
/// its path. The filename includes the PID so concurrent processes don't clash.
fn write_bridge_script() -> io::Result<PathBuf> {
    let mut dir = std::env::temp_dir();
    dir.push(format!("tuxlink-catptt-bridge-{}.py", std::process::id()));
    let mut f = std::fs::File::create(&dir)?;
    f.write_all(BRIDGE_SCRIPT.as_bytes())?;
    f.flush()?;
    Ok(dir)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn spec() -> CatBridgeSpec {
        CatBridgeSpec {
            bridge_port: 4532,
            serial_path: "/dev/ttyUSB0".into(),
            baud: 38400,
            key_cmd: "TX1;".into(),
            unkey_cmd: "TX0;".into(),
        }
    }

    #[test]
    fn python_args_carries_every_parameter() {
        let args = spec().python_args("/tmp/bridge.py");
        assert_eq!(args[0], "/tmp/bridge.py");
        // Each flag/value pair must be present in order.
        assert!(args.windows(2).any(|w| w[0] == "--port" && w[1] == "4532"), "{args:?}");
        assert!(args.windows(2).any(|w| w[0] == "--serial" && w[1] == "/dev/ttyUSB0"), "{args:?}");
        assert!(args.windows(2).any(|w| w[0] == "--baud" && w[1] == "38400"), "{args:?}");
        assert!(args.windows(2).any(|w| w[0] == "--key" && w[1] == "TX1;"), "{args:?}");
        assert!(args.windows(2).any(|w| w[0] == "--unkey" && w[1] == "TX0;"), "{args:?}");
    }

    #[test]
    fn python_args_honors_custom_parameters() {
        let s = CatBridgeSpec {
            bridge_port: 4600,
            serial_path: "/dev/ttyUSB2".into(),
            baud: 9600,
            key_cmd: "RT1;".into(),
            unkey_cmd: "RT0;".into(),
        };
        let args = s.python_args("/tmp/b.py");
        assert!(args.windows(2).any(|w| w[0] == "--port" && w[1] == "4600"), "{args:?}");
        assert!(args.windows(2).any(|w| w[0] == "--serial" && w[1] == "/dev/ttyUSB2"), "{args:?}");
        assert!(args.windows(2).any(|w| w[0] == "--baud" && w[1] == "9600"), "{args:?}");
        assert!(args.windows(2).any(|w| w[0] == "--key" && w[1] == "RT1;"), "{args:?}");
        assert!(args.windows(2).any(|w| w[0] == "--unkey" && w[1] == "RT0;"), "{args:?}");
    }

    #[test]
    fn embedded_script_is_the_parameterized_bridge() {
        // Guard against the embed pointing at a stale/hardcoded script: the
        // parameterized bridge MUST parse argv and send the unkey failsafe.
        assert!(BRIDGE_SCRIPT.contains("argparse"), "bridge script must parse argv");
        assert!(BRIDGE_SCRIPT.contains("--unkey"), "bridge script must accept --unkey");
        assert!(
            BRIDGE_SCRIPT.contains("failsafe"),
            "bridge script must send the unkey failsafe on teardown"
        );
    }

    #[test]
    fn write_bridge_script_writes_the_embedded_content() {
        let path = write_bridge_script().expect("write temp script");
        let content = std::fs::read_to_string(&path).expect("read back");
        assert_eq!(content, BRIDGE_SCRIPT);
        let _ = std::fs::remove_file(&path);
    }
}
